llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -761,5 +761,147 @@ getLargeGRFSplitter(ModuleDesc &&MD, bool EmitOnlyKernelsAsEntryPoints) {
     return std::make_unique<ModuleCopier>(std::move(MD), std::move(Groups));
 }
 
+namespace {
+// Data structure, which represent a combination of all possible optional
+// features used in a function.
+//
+// It has extra methods to be useable as a key in llvm::DenseMap.
+struct UsedOptionalFeatures {
+  SmallVector<int, 4> Aspects;
+  // TODO: extend this further with reqd-sub-group-size, reqd-work-group-size,
+  // double-grf and other properties
```

Comment:
They are applied, but if we have two kernels with different required sub-group sizes, they must be split into a separate modules or otherwise JIT compilation might fail, because not all features are required to be supported by all devices for all sub-group sizes. The same is applicable for required work-group size, because different devices might have different supported work-group sizes.